### PR TITLE
support deep object validate in form

### PR DESCRIPTION
**Pull request by baiyaaaaa**
Please makes sure these boxes are checked before submitting your PR, thank you!

* [ ] Make sure you follow Element's [Contributing Guide](https://github.com/ElemeFE/element/blob/master/.github/CONTRIBUTING.md).
* [ ] Make sure you are merging your commits to `dev` branch.
* [ ] Rebase before creating a PR to keep commit history clear.
* [ ] Add some descriptions and refer relative issues for you PR.

### CODE REVIEW

**Comment by baiyaaaaa** [comment]
#1328

File: packages/form/src/form.vue [hunk 1]
```diff
@@ -25,49 +25,46 @@
...
+        if (field) {
+          this.fields.push(field);
+        }
       });
       /* istanbul ignore next */
       this.$on('el.form.removeField', (field) => {
-        if (this.fields[field.prop]) {
-          delete this.fields[field.prop];
-          this.fieldLength--;
+        console.log(field);
```
Comment (QingWei-Li) [inline packages/form/src/form.vue]: ...

File: packages/form/src/form.vue [hunk 2]
```diff
@@ -25,49 +25,46 @@
...
+        if (field.prop) {
+          this.fields.splice(this.fields.indexOf(field), 1);
```
Comment (QingWei-Li) [inline packages/form/src/form.vue]: this.fields.indexOf(field) 如果为 -1 会误删掉最后一个

File: packages/form/src/form.vue [hunk 3]
```diff
@@ -25,49 +25,46 @@
...
               callback(valid);
             }
           });
-        }
+        });
       },
       validateField(prop, cb) {
-        var field = this.fields[prop];
+        var field = this.fields.filter(field => {
+          return field.prop === prop;
```
Comment (QingWei-Li) [inline packages/form/src/form.vue]: 可以写成一行了

File: examples/docs/en-US/form.md [hunk 4]
```diff
@@ -782,7 +786,7 @@ Form component allows you to verify your data, helping you find and correct erro
       },
       addDomain() {
         this.dynamicForm.domains.push({
-          key: this.dynamicForm.domains.length,
+          key: Date.now,
```
Comment (Leopoldthecoder) [inline examples/docs/en-US/form.md]: `Date.now()`

File: packages/form/src/form.vue (hunk 1 above)
Comment (Leopoldthecoder) [inline packages/form/src/form.vue]: 🌚

Comment (Kingwl) [inline packages/form/src/form.vue]: 2333333

**Comment by baiyaaaaa** [comment]
#1077

**Pull request merged**